src/mongocxx/test/spec/unified_test_format.cpp

```diff
@@ -113,6 +111,180 @@ bool has_run_on_requirements(const bsoncxx::document::view test) {
+
+    using namespace std;
+    auto should_remove = [&](const char c) {
+        const auto remove = {' ', '"', '{', '}'};
+        return end(remove) != find(begin(remove), end(remove), c);
+    };
+
+    s.erase(remove_if(begin(s), end(s), should_remove), end(s));
+    replace(begin(s), end(s), ':', '=');
+    return s;
+}
+
+std::string json_to_uri_opts(const std::string& input) {
+    // Transforms a non-nested JSON document string (assumed to contain URI keys and values) to a
+    // string of equivalent URI options and values. That is,
+    //      input   := "{ "readConcernLevel" : "local", "w" : 1 }"
+    //      output  := "readConcernLevel=local&w=1"
+    std::vector<std::string> output;
+    const std::regex delim(",");
+    std::transform(std::sregex_token_iterator(std::begin(input), std::end(input), delim, -1),
+                   std::sregex_token_iterator(),
+                   std::back_inserter(output),
+                   json_kvp_to_uri_kvp);
+
+    auto join = [](const std::string& s1, const std::string& s2) { return s1 + "&" + s2; };
+    return std::accumulate(std::begin(output) + 1, std::end(output), output[0], join);
+}
+
+std::string uri_options_to_string(document::view object) {
+    // Spec: Optional object. Additional URI options to apply to the test suite's connection string
+    // that is used to create this client. Any keys in this object MUST override conflicting keys in
+    // the connection string.
+    if (!object["uriOptions"])
+        return {};
+
+    // TODO: Spec: if 'readPreferenceTags' is specified in this object, the key will map to an array
+    //  of strings, each representing a tag set, since it is not feasible to define multiple
+    //  'readPreferenceTags' keys in the object.
+    REQUIRE_FALSE(object["readPreferenceTags"]);
+
+    auto json = to_json(object["uriOptions"].get_document());
+    auto opts = json_to_uri_opts(json);
+
+    CAPTURE(json, opts);
+    return opts;
+}
+
+std::string get_hostnames(document::view object) {
+    auto default_uri{"localhost:27017"};
+    // Spec: This [useMultipleMongoses] option has no effect for non-sharded topologies.
+    if (test_util::is_replica_set())
+        return default_uri;
+
+    // Spec: If true and the topology is a sharded cluster, the test runner MUST assert that this
+    // MongoClient connects to multiple mongos hosts (e.g. by inspecting the connection string).
+    if (!object["useMultipleMongoses"] || !object["useMultipleMongoses"].get_bool())
+        return default_uri;
+
+    // from: https://docs.mongodb.com/manual/reference/config-database/#config.shards
+    // If the shard is a replica set, the host field displays the name of the replica set, then a
+    // slash, then a comma-separated list of the hostnames of each member of the replica set, as in
+    // the following example:
+    //      { ... , "host" : "shard0001/localhost:27018,localhost:27019,localhost:27020", ... }
```

Comment:
Oh good idea. C currently hard codes `localhost:27018` for `useMultipleMongoses`. I noted this in CDRIVER-3285.